### PR TITLE
Fix marketplace promotions action scheduler

### PR DESCRIPTION
**Pull request by andfinally**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

WooCommerce 8.8.1 introduced some code dependent on Action Scheduler 3.3+. Under normal circumstances that'd be ok, as this WooCommerce version comes bundled with A-S 3.7.0. But other plugins that also bundle A-S might be loading their version of A-S unusually early, preventing the one in WC from loading. In this situation, activating WooCommerce would give rise to this fatal error.

```
PHP Fatal error:  Uncaught Error: Call to undefined function as_has_scheduled_action() in /wordpress/plugins/woocommerce/8.8.1/includes/admin/class-wc-admin-marketplace-promotions.php:80
Stack trace:
#0 /wordpress/core/6.5.2/wp-includes/class-wp-hook.php(324): WC_Admin_Marketplace_Promotions::schedule_promotion_fetch('')
#1 /wordpress/core/6.5.2/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters(NULL, Array)
#2 /wordpress/core/6.5.2/wp-includes/plugin.php(517): WP_Hook->do_action(Array)
#3 /wordpress/core/6.5.2/wp-settings.php(695): do_action('init')
#4 /srv/htdocs/wp-config.php(81): require_once('/wordpress/core...')
#5 /wordpress/core/6.5.2/wp-load.php(55): require_once('/srv/htdocs/wp-...')
#6 /wordpress/core/6.5.2/wp-admin/admin-ajax.php(22): require_once('/wordpress/core...')
#7 {main}
  thrown in /wordpress/plugins/woocommerce/8.8.1/includes/admin/class-wc-admin-marketplace-promotions.php on line 80
```

This PR adds defensive checks to `WC_Admin_Marketplace_Promotions` to ensure A-S functions are available before it calls them.

### Notes

- Other parts of WooCommerce also call `as_has_scheduled_action`: [BatchProcessingController](https://github.com/woocommerce/woocommerce/blob/c752c60fd40fdc976381fcf9c8b248f2e3666999/plugins/woocommerce/src/Internal/BatchProcessing/BatchProcessingController.php#L114) and [TransientFilesEngine](https://github.com/woocommerce/woocommerce/blob/eb9723e5e0b3b1b2000f21ec335823e87ee31ed1/plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php#L317). However, these other calls do not cause errors which totally break wp-admin. This critical error seems more urgent, so I haven't touched those other areas in this PR. I've created a separate issue for them: https://github.com/woocommerce/woocommerce/issues/46632.
- I'm currently seeing an issue on my local wp-env environment on the marketplace Discover tab http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=extensions&path=%2Fextensions. The page's request to https://woocommerce.test/wp-json/wccom-extensions/2.0/featured is returning `cURL error 35: OpenSSL SSL_connect: SSL_ERROR_SYSCALL in connection to [woocommerce.com:443](http://woocommerce.com:443/)`. So the content isn't rendered. This is not related to the current change: it happens in trunk too. It's probably due to local configuration on wp-env.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Disable WooCommerce in your local testing environment.
2. Check out this branch. Build WooCommerce:

```bash
nvm use
pnpm install
pnpm build
```

3. Install and activate WP Mail SMTP 2.8.0. You can use this file: [wp-mail-smtp.2.8.0.zip](https://github.com/woocommerce/woocommerce/files/14994676/wp-mail-smtp.2.8.0.zip). 
4. Activate WooCommerce. You should be able to view wp-admin without getting the fatal error.

#### Testing the marketplace promotions still work

5. Deactivate WP Mail SMTP.
6. Edit `plugins/woocommerce/includes/admin/class-wc-admin-marketplace-promotions.php` and replace line 97 with this. This makes a request to a gist with dummy data that is valid until 1 May.

```php
$raw_promotions = wp_safe_remote_get( 'https://gist.githubusercontent.com/andfinally/acd8646151c0108f92979c985c20a0d8/raw/promotions.json' );
```
7. On line 21, change the frequency of the scheduled action to five minutes:

```php
const SCHEDULED_ACTION_INTERVAL = 5 * MINUTE_IN_SECONDS;
```

8. Visit the [pending scheduled actions page](http://localhost:8888/wp-admin/admin.php?page=wc-status&tab=action-scheduler&status=pending) and search for `woocommerce_marketplace_fetch_promotions`. If there's a pending action, cancel it using the "cancel" hover link under it in the list. A new action should be scheduled.
9. After 5 minutes or so, the action should run. (You may need to jog it by viewing a page on your test site.)
10. The menu bubble should appear on the `WooCommerce > Extensions` menu item, and the notice should appear on Marketplace pages like `http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fextensions`.

<p>
image
</p>

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Review by corsonr — APPROVED** [review]
All good, I can confirm that there's no more fatal error, and that the promotions do show fine on the extensions page so as the menu bubble.

**Note to other reviewers:** the gist adjustments should be done at line 100, not 97. Also do not forget to revert changes and cancel the 5min `woocommerce_marketplace_fetch_promotions` scheduled action so that it goes back to 12h.

**Comment by brettnzl** [comment]
@corsonr this fix has caused an issue with WooCommerce "Woo Subscriptions" from the day we updated to this, we have been getting several "Failed-scheduled-actions" logs. 

Screenshot 2024-04-22 090402

2024-04-21T00:08:41+00:00 Notice scheduled action 665467 (subscription payment) failed to finish processing due to the following exception: Call to a member function meta_exists() on null  
Additional context
{
    "_legacy": true
}
2024-04-21T00:08:41+00:00 Notice action args: subscription_id: 7212  
Additional context
{
    "_legacy": true
}
2024-04-21T00:56:42+00:00 Notice scheduled action 665567 (subscription payment) failed to finish processing due to the following exception: Call to a member function meta_exists() on null  
Additional context
2024-04-21T00:56:42+00:00 Notice action args: subscription_id: 9027  
Additional context

**Comment by corsonr** [comment]
A fix has already been pushed by @andfinally if I am correct
